guacamole-ext/src/main/java/org/apache/guacamole/net/auth/Credentials.java

```diff
@@ -98,20 +104,43 @@ public class Credentials implements Serializable {
      *     The HTTP request associated with the authentication
      *     request.
      */
-    public Credentials(String username, String password, HttpServletRequest request) {
+    public Credentials(String username, String password, 
+            HttpServletRequest request) {
         this.username = username;
         this.password = password;
         this.request = request;
+        this.authenticationResumed = false;
+        
+        if (request != null) {
+            // Set the remote address
+            this.remoteAddress = request.getRemoteAddr();
 
-        // Set the remote address
-        this.remoteAddress = request.getRemoteAddr();
+            // Get the remote hostname
+            this.remoteHostname = request.getRemoteHost();
 
-        // Get the remote hostname
-        this.remoteHostname = request.getRemoteHost();
+            // If session exists get it, but don't create a new one.
+            this.session = request.getSession(false);
+        }
 
-        // If session exists get it, but don't create a new one.
-        this.session = request.getSession(false);
+    }
+
+    /**
+     * Checks if the current authentication process is a resumed one.
+     *
+     * @return True if authentication is resumed, otherwise false.
+     */
+    public Boolean isAuthenticationResumed() {
```

Comment:
It'd be nice if this could avoid going in `Credentials`, because it's pretty specific to Duo, and it's also not even the only way of resuming sessions - see the SSO (SAML, etc) extensions, which have their own [completely-different mechanism](https://github.com/apache/guacamole-client/blob/main/extensions/guacamole-auth-sso/modules/guacamole-auth-sso-base/src/main/java/org/apache/guacamole/auth/sso/SSOAuthenticationEventListener.java) for resuming interrupted auth attempts. 

Maybe the DUO extension could do something similar? (using a URL parameter to track state). If there could be some shared functionality for this, it would be ideal, though that seems unlikely considering how different the implementations are.

At the bare minimum, I think that the SSO extensions should at least _set_ this flag.

Reply:
The idea was to migrate sso to this method

Reply:
It looks like the DUO extension is already[ tracking state through URL parameters](https://github.com/apache/guacamole-client/blob/0e12f48c29bb872b2b73873427430286aa27202d/extensions/guacamole-auth-duo/src/main/java/org/apache/guacamole/auth/duo/UserVerificationService.java#L123).

Reply:
> The query strings are defined by the authentication provider modules so that we can avoid accidental overlap

FWIW, the ULR parameter for SAML is also called "state". Likely things won't work if people try to use DUO + SAML together. Might want to consider changing one or both of those.

Were you planning on migrating the SSO extensions to use this new flag in this PR?

Reply:
You're correct that the SAML extension also uses a `state` parameter, which could lead to confusion when both `DUO` and `SAML` are used concurrently, but I also added the `provider_id` as a required query parameter for resuming authentication. This acts as a namespacing mechanism to prevent collision. When redirecting from duo the query parameters will look like such: `provider_id=duo&state=abc123&duo_code=abc123`. 

This PR isn't meant for migrating the SSO extensions but their was the plan to eventually migrate them to use the same code for resuming authentication.

Reply:
It definitely feels like we shouldn't have the SSO extensions unmigrated for very long ... what do you think @mike-jumper ? Should migration of the SSO extensions to this new system be part of this ticket?

Reply:
I would agree that the SSO extension migration should be with this ticket - either this PR, or another PR to follow quickly after this one...